plugins/GSdx/res/tfx.fx

```diff
@@ -871,6 +881,40 @@ void gs_main(line VS_OUTPUT input[2], inout TriangleStream<VS_OUTPUT> stream)
 
 #endif
 
+void ps_blend(inout float4 Color, float As, float2 pos_xy)
+{
+	if (SW_BLEND)
+	{
+		float4 RT = RtSampler.Load(int3(pos_xy, 0));
+
+		float3 Cs = trunc(Color.rgb * 255.0f + 0.1f);
+		float3 Cd = trunc(RT.rgb * 255.0f + 0.1f);
+		float3 Cv;
+
+		float Ad = (PS_DFMT == FMT_24) ? 1.0f : (RT.a * 255.0f / 128.0f);
+
+		float3 A = (PS_BLEND_A == 0) ? Cs : ((PS_BLEND_A == 1) ? Cd : (float3)0.0f);
+		float3 B = (PS_BLEND_B == 0) ? Cs : ((PS_BLEND_B == 1) ? Cd : (float3)0.0f);
+		float3 C = (PS_BLEND_C == 0) ? As : ((PS_BLEND_C == 1) ? Ad : Af);
+		float3 D = (PS_BLEND_D == 0) ? Cs : ((PS_BLEND_D == 1) ? Cd : (float3)0.0f);
+
+		Cv = (PS_BLEND_A == PS_BLEND_B) ? D : trunc(((A - B) * C) + D);
+
+		// Standard Clamp
+		if (PS_COLCLIP == 0 && PS_HDR == 0)
+			Cv = clamp(Cv, (float3)0.0f, (float3)255.0f);
+
+		// In 16 bits format, only 5 bits of color are used. It impacts shadows computation of Castlevania
+		// FIXME: Should Alpha be masked also here or not?
+		if (PS_DFMT == FMT_16)
+			Cv = (float3)((uint3)Cv & (uint3)0xF8);
+		else
```

Comment:
Maybe `else if PS_COLCLIP == 1 && PS_HDR == 0` ? like on gl? I guess Greg will tell us which is better.

Reply:
I saw those checks in the OGL shader but I don't know if is totally necessary. If doing clamping, it is already done before the masking so masking will not change the result. At worst it will just be doing an unnecessary mask. In the HDR case (not clamping) it will not affect the end result since we will be throwing away the upper bits in the final pass (in 16 bit format this is done anyway even in OGL).

I think it is clearer like this that the HDR algo is not affected by the masking, especially if something were to change with the HDR algo later on. Maybe there is something else I am not understanding here like maybe the perf impact of converting to int => masking => convert back to float. Like you said it's probably best to wait for gregory's input here.